### ISSUE TITLE
Dataframe does not support duplicate column names perfectly

### ISSUE THREAD

**Issue by datajanko** [dask/dask#4079]
I just realized that dask datafrmaes, in contrast to pandas, does not support duplicate column names (perfectly).

```
import dask.array as da
import dask.dataframe as dd
import numpy as np
arr = da.from_array(np.arange(10).reshape(5,2), chunks=(5,2))
dd.from_dask_array(arr, columns=['a', 'a'])
```
raises the following error message:

<details>

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/anaconda3/envs/dask-ml-dev/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/anaconda3/envs/dask-ml-dev/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    398                         if cls is not object \
    399                                 and callable(cls.__dict__.get('__repr__')):
--> 400                             return _repr_pprint(obj, self, cycle)
    401 
    402             return _default_pprint(obj, self, cycle)

~/anaconda3/envs/dask-ml-dev/lib/python3.6/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    693     """A pprint that just redirects to the normal repr function."""
    694     # Find newlines and replace them with p.break_()
--> 695     output = repr(obj)
    696     for idx,output_line in enumerate(output.splitlines()):
    697         if idx:

~/anaconda3/envs/dask-ml-dev/lib/python3.6/site-packages/dask/dataframe/core.py in __repr__(self)
    392 
    393     def __repr__(self):
--> 394         data = self._repr_data.to_string(max_rows=5, show_dimensions=False)
    395         return """Dask {klass} Structure:
    396 {data}

~/anaconda3/envs/dask-ml-dev/lib/python3.6/site-packages/dask/dataframe/core.py in __getattr__(self, key)
   2518             return new_dd_object(merge(self.dask, dsk), name,
   2519                                  meta, self.divisions)
-> 2520         raise AttributeError("'DataFrame' object has no attribute %r" % key)
   2521 
   2522     def __dir__(self):

AttributeError: 'DataFrame' object has no attribute '_repr_data'

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/anaconda3/envs/dask-ml-dev/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    343             method = get_real_method(obj, self.print_method)
    344             if method is not None:
--> 345                 return method()
    346             return None
    347         else:

~/anaconda3/envs/dask-ml-dev/lib/python3.6/site-packages/dask/dataframe/core.py in _repr_html_(self)
   3131 
   3132     def _repr_html_(self):
-> 3133         data = self._repr_data.to_html(max_rows=5,
   3134                                        show_dimensions=False, notebook=True)
   3135         return self._HTML_FMT.format(data=data, name=key_split(self._name),

~/anaconda3/envs/dask-ml-dev/lib/python3.6/site-packages/dask/dataframe/core.py in __getattr__(self, key)
   2518             return new_dd_object(merge(self.dask, dsk), name,
   2519                                  meta, self.divisions)
-> 2520         raise AttributeError("'DataFrame' object has no attribute %r" % key)
   2521 
   2522     def __dir__(self):

AttributeError: 'DataFrame' object has no attribute '_repr_data'
```

</details>

However, assigning the frame to a variable and computing the array works
```
m = dd.from_dask_array(arr, columns=['a', 'a'])
m.compute()

  | a | a
-- | -- | --
0 | 1
2 | 3
4 | 5
6 | 7
8 | 9

```

I didn't find the issue, so please apologies if it's a duplicate

**Comment by mrocklin** [comment]
Thanks for the bug report with a great example.  It looks like this actually create the dask dataframe, but that it fails when repring the result.  

I think that it fails here: 

https://github.com/dask/dask/blob/master/dask/dataframe/core.py#L3123-L3128

In particular on this line:

https://github.com/dask/dask/blob/master/dask/dataframe/core.py#L3127

Because now `meta[c]` is actually a Pandas DataFrame rather than a Pandas Series as we might expect.  Once we get past this (which shouldn't be too hard) we would also have to stop using a dictionary and start using a list of tuples or something else that would support repeats.

@datajanko do you have any interest in diving in here?  This is mostly an issue about manipulating core pandas objects rather than any dask fanciness.

**Comment by datajanko** [comment]
Yes, I'd like to dive in there, maybe even before completing my dask_ml task(s).

**Comment by jai-soni** [comment]
I am still facing this issue in dask 2.3.0. Any resolution to this?
Gives me error like:
`---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-c6ddddbf1a1b> in <module>
      1 t0 = time.time()
      2 result = dask_dataframe.map_partitions(getLower, meta=data_dask)
----> 3 data_dask = result.compute()
      4 t1 = time.time()
      5 print("Time to process with Dask {}".format(t1-t0))

C:\ProgramData\Anaconda3\lib\site-packages\dask\base.py in compute(self, **kwargs)
    173         dask.base.compute
    174         """
--> 175         (result,) = compute(self, traverse=False, **kwargs)
    176         return result
    177 

C:\ProgramData\Anaconda3\lib\site-packages\dask\base.py in compute(*args, **kwargs)
    444     keys = [x.__dask_keys__() for x in collections]
    445     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 446     results = schedule(dsk, keys, **kwargs)
    447     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    448 

C:\ProgramData\Anaconda3\lib\site-packages\dask\threaded.py in get(dsk, result, cache, num_workers, pool, **kwargs)
     80         get_id=_thread_get_id,
     81         pack_exception=pack_exception,
---> 82         **kwargs
     83     )
     84 

C:\ProgramData\Anaconda3\lib\site-packages\dask\local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    489                         _execute_task(task, data)  # Re-execute locally
    490                     else:
--> 491                         raise_exception(exc, tb)
    492                 res, worker_id = loads(res_info)
    493                 state["cache"][key] = res

C:\ProgramData\Anaconda3\lib\site-packages\dask\compatibility.py in reraise(exc, tb)
    128         if exc.__traceback__ is not tb:
    129             raise exc.with_traceback(tb)
--> 130         raise exc
    131 
    132     import pickle as cPickle

C:\ProgramData\Anaconda3\lib\site-packages\dask\local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    231     try:
    232         task, data = loads(task_info)
--> 233         result = _execute_task(task, data)
    234         id = get_id()
    235         result = dumps((result, id))

C:\ProgramData\Anaconda3\lib\site-packages\dask\core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

C:\ProgramData\Anaconda3\lib\site-packages\dask\optimization.py in __call__(self, *args)
   1057         if not len(args) == len(self.inkeys):
   1058             raise ValueError("Expected %d args, got %d" % (len(self.inkeys), len(args)))
-> 1059         return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
   1060 
   1061     def __reduce__(self):

C:\ProgramData\Anaconda3\lib\site-packages\dask\core.py in get(dsk, out, cache)
    147     for key in toposort(dsk):
    148         task = dsk[key]
--> 149         result = _execute_task(task, cache)
    150         cache[key] = result
    151     result = _execute_task(out, cache)

C:\ProgramData\Anaconda3\lib\site-packages\dask\core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

C:\ProgramData\Anaconda3\lib\site-packages\dask\compatibility.py in apply(func, args, kwargs)
    105     def apply(func, args, kwargs=None):
    106         if kwargs:
--> 107             return func(*args, **kwargs)
    108         else:
    109             return func(*args)

C:\ProgramData\Anaconda3\lib\site-packages\dask\dataframe\core.py in apply_and_enforce(*args, **kwargs)
   4857                 c = meta.columns
   4858         else:
-> 4859             c = meta.name
   4860         return _rename(c, df)
   4861     return df

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\generic.py in __getattr__(self, name)
   5065             if self._info_axis._can_hold_identifiers_and_holds_name(name):
   5066                 return self[name]
-> 5067             return object.__getattribute__(self, name)
   5068 
   5069     def __setattr__(self, name, value):

AttributeError: 'DataFrame' object has no attribute 'name'`

**Comment by datajanko** [comment]
could you provide more context? maybe your input? are you sure this is related to this issue

**Comment by jai-soni** [comment]
I think its the issue with the input data. I was trying out dask to see if it improves performance for my text processing task so I created a simple function that returns text in lowecase. My getLower function ingested pandas df and returned `df.map(lambda x: x.lower())` which resulted in the above error. The only change I did to make it work was to change getLower to returning `df.apply(lambda x : x.lower())` and I don't get this error. Seeing the error message in the comments earlier, I think its a similar issue.
